topdown/object_test.go

```diff
@@ -0,0 +1,262 @@
+		object   string
+		key      interface{}
+		fallback interface{}
+		expected interface{}
+	}{
+		{
+			note:     "basic case . found",
+			object:   `{"a": "b"}`,
+			key:      `"a"`,
+			fallback: `"c"`,
+			expected: `"b"`,
+		},
+		{
+			note:     "basic case . not found",
+			object:   `{"a": "b"}`,
+			key:      `"c"`,
+			fallback: `"c"`,
+			expected: `"c"`,
+		},
+		{
+
+			note:     "integer key . found",
+			object:   "{1: 2}",
+			key:      "1",
+			fallback: "3",
+			expected: "2",
+		},
+		{
+			note:     "integer key . not found",
+			object:   "{1: 2}",
+			key:      "2",
+			fallback: "3",
+			expected: "3",
+		},
+		{
+			note:     "complex value . found",
+			object:   `{"a": {"b": "c"}}`,
+			key:      `"a"`,
+			fallback: "true",
+			expected: `{"b": "c"}`,
+		},
+		{
+			note:     "complex value . not found",
+			object:   `{"a": {"b": "c"}}`,
+			key:      `"b"`,
+			fallback: "true",
+			expected: "true",
+		},
+	}
+
+	for _, tc := range cases {
+		rules := []string{
+			fmt.Sprintf("p = x { x := object.get(%s, %s, %s) }", tc.object, tc.key, tc.fallback),
+		}
+		runTopDownTestCase(t, map[string]interface{}{}, tc.note, rules, tc.expected)
+	}
+}
+
+func TestBuiltinObjectUnion(t *testing.T) {
+	cases := []struct {
+		note     string
+		objectA  string
+		objectB  string
+		expected interface{}
+	}{
+		{
+			note:     "both empty",
+			objectA:  `{}`,
+			objectB:  `{}`,
+			expected: `{}`,
+		},
+		{
+			note:     "left empty",
+			objectA:  `{}`,
+			objectB:  `{"a": 1}`,
+			expected: `{"a": 1}`,
+		},
+		{
+			note:     "right empty",
+			objectA:  `{"a": 1}`,
+			objectB:  `{}`,
+			expected: `{"a": 1}`,
+		},
+		{
+			note:     "base",
+			objectA:  `{"a": 1}`,
+			objectB:  `{"b": 2}`,
+			expected: `{"a": 1, "b": 2}`,
+		},
+		{
+			note:     "nested",
+			objectA:  `{"a": {"b": {"c": 1}}}`,
+			objectB:  `{"b": 2}`,
+			expected: `{"a": {"b": {"c": 1}}, "b": 2}`,
+		},
+		{
+			note:     "nested reverse",
+			objectA:  `{"b": 2}`,
+			objectB:  `{"a": {"b": {"c": 1}}}`,
+			expected: `{"a": {"b": {"c": 1}}, "b": 2}`,
+		},
+		{
+			note:     "conflict simple",
+			objectA:  `{"a": 1}`,
+			objectB:  `{"a": 2}`,
+			expected: `{"a": 1}`,
+		},
+		{
+			note:     "conflict nested and extra field",
+			objectA:  `{"a": 1}`,
+			objectB:  `{"a": {"b": {"c": 1}}, "d": 7}`,
+			expected: `{"a": 1, "d": 7}`,
+		},
+		{
+			note:     "conflict multiple",
+			objectA:  `{"a": {"b": {"c": 1}}, "e": 1}`,
+			objectB:  `{"a": {"b": "foo", "b1": "bar"}, "d": 7, "e": 17}`,
+			expected: `{"a": {"b": {"c": 1}, "b1": "bar"}, "d": 7, "e": 1}`,
+		},
+	}
+
+	for _, tc := range cases {
+		rules := []string{
+			fmt.Sprintf("p = x { x := object.union(%s, %s) }", tc.objectA, tc.objectB),
+		}
+		runTopDownTestCase(t, map[string]interface{}{}, tc.note, rules, tc.expected)
+	}
+}
+
+func TestBuiltinObjectRemove(t *testing.T) {
+	cases := []struct {
+		note     string
+		object   string
+		keys     string
+		expected interface{}
```

Comment:
I noticed we don't have any error cases.

Reply:
I remembered why this was like this... the only errors are type-related which are caught by the compiler. The topdown helper wasn't really able to help validate them (and most builtins are guilty of this type of happy path testing WRT param types errors).

Some tweaks to the topdown helpers and we're in business now..

Reply:
The compiler can't catch type errors for data for which it can't infer the types though (e.g., `input`). We should be testing error paths otherwise we'll end up with undefined behaviour (e.g., panics).